test/util.sh

```diff
@@ -0,0 +1,203 @@
+	cmd=$1
+
+	os::test::expect_exit_code_run_grep "${cmd}"
+}
+
+# expect_failure runs the cmd and expects a non-zero exit code
+function os::test::expect_failure() {
+	cmd=$1
+
+	os::test::expect_exit_code_run_grep "${cmd}" "failure_func"
+}
+
+# expect_success_and_text runs the cmd and expects an exit code of 0
+# as well as running a grep test to find the given string in the output
+function os::test::expect_success_and_text() {
+	cmd=$1
+	expected_text=$2
+
+	os::test::expect_exit_code_run_grep "${cmd}" "success_func" "${expected_text}"
+}
+
+# expect_failure_and_text runs the cmd and expects a non-zero exit code
+# as well as running a grep test to find the given string in the output
+function os::test::expect_failure_and_text() {
+	cmd=$1
+	expected_text=$2
+
+	os::test::expect_exit_code_run_grep "${cmd}" "failure_func" "${expected_text}"
+}
+
+# expect_success_and_not_text runs the cmd and expects an exit code of 0
+# as well as running a grep test to ensure the given string is not in the output
+function os::test::expect_success_and_not_text() {
+	cmd=$1
+	expected_text=$2
+
+	os::test::expect_exit_code_run_grep "${cmd}" "success_func" "${expected_text}" "failure_func"
+}
+
+# expect_failure_and_not_text runs the cmd and expects a non-zero exit code
+# as well as running a grep test to ensure the given string is not in the output
+function os::test::expect_failure_and_not_text() {
+	cmd=$1
+	expected_text=$2
+
+	os::test::expect_exit_code_run_grep "${cmd}" "failure_func" "${expected_text}" "failure_func"
+}
+
+# MEDIUM-LEVEL WRAPPER FUNCTIONS: ASSUME NO CODES
+
+# expect_code runs the cmd and expects a given exit code
+function os::test::expect_code() {
+	cmd=$1
+	expected_cmd_code=$2
+
+	os::test::expect_exit_code_run_grep "${cmd}" "specific_code_func ${expected_cmd_code}"
+}
+
+# expect_code_and_text runs the cmd and expects the given exit code
+# as well as running a grep test to find the given string in the output
+function os::test::expect_code_and_text() {
+	cmd=$1
+	expected_cmd_code=$2
+	expected_text=$3
+
+	os::test::expect_exit_code_run_grep "${cmd}" "specific_code_func ${expected_cmd_code}" "${expected_text}"
+}
+
+# expect_code_and_not_text runs the cmd and expects the given exit code
+# as well as running a grep test to ensure the given string is not in the output
+function os::test::expect_code_and_not_text() {
+	cmd=$1
+	expected_cmd_code=$2
+	expected_text=$3
+
+	os::test::expect_exit_code_run_grep "${cmd}" "specific_code_func ${expected_cmd_code}" "${expected_text}" "failure_func"
+}
+
+# TEST EXECUTION FUNCTION
+
+# expect_exit_code_and_text runs the provided test command and expects a specific exit code from that
+# command as well as the success of a specified `grep` invocation. Output from the command to be 
+# tested is suppressed unless either `VERBOSE=1` or the test fails.
+function os::test::expect_exit_code_run_grep() {
+	cmd=$1
+	# default expected cmd code to 0 for success
+	cmd_eval_func=${2:-success_func}
```

Comment:
> Using function pointers to implement a strategy pattern is not "really ugly". Please elaborate on why you think this is worse than passing in boolean (which type? Those that work with []? Or those that are human readable but don't?) flags and doing a long if/else if chain.

With a type safe language that doesn't allow arbitrary execution of typos, its decidedly less ugly.  In bash, this is error prone and dangerous.  Even in a language that has reasonable support for making it work, requiring strategy arguments like this for flexibility that isn't required isn't proper.  We don't care about what a failing exit code's value is, only that it failed: this is flexibility that doesn't matter.